packages/ai/src/ui/chat-store.ts

```diff
@@ -0,0 +1,518 @@
+          activeResponse,
+          chatId,
+        });
+        break;
+      }
+      // Parts that are just appended to parts array:
+      case 'step-start':
+      case 'source':
+      case 'file': {
+        activeResponse.message.parts.push(partDelta);
+        break;
+      }
+      default: {
+        throw new Error('Invalid message part type');
+      }
+    }
+
+    chat.activeResponse!.message.revisionId = generateId();
+    chat.messages = [...chat.messages];
+    this.emitEvent({ id: chatId, event: ChatStoreEvent.ChatMessagesChanged });
+  }
+
+  updateActiveResponse({
+    chatId,
+    message,
+    generateId,
+  }: {
+    chatId: string;
+    message: Partial<UIMessage>;
+    generateId: () => string;
+  }) {
+    const chat = this.chats.get(chatId);
+    if (!chat) return;
+
+    if (!chat.activeResponse) {
+      this.initializeActiveResponse({
+        chatId,
+        messageId: message.id,
+        generateId,
+      });
+    }
+
+    chat.activeResponse!.message = {
+      ...chat.activeResponse!.message,
+      ...message,
+    };
+
+    this.setMessages({
+      id: chatId,
+      messages: [...chat.messages.slice(0, -1), chat.activeResponse!.message],
+    });
+  }
+
+  clearStepPartialState({
+    id,
+    isContinued = false,
+  }: {
+    id: string;
+    isContinued?: boolean;
+  }) {
+    const chat = this.chats.get(id);
+    if (!chat || !chat.activeResponse) return;
+
+    const partialState = {
+      toolParts: chat.activeResponse?.partialState?.toolParts,
+      reasoningPart: undefined,
+      textPart: isContinued
+        ? chat.activeResponse?.partialState?.textPart
+        : undefined,
+    };
+
+    chat.activeResponse.partialState = partialState;
+  }
+
+  private async addOrUpdateToolInvocation({
+    chatId,
+    toolInvocation,
+    activeResponse,
+  }: {
+    chatId: string;
+    toolInvocation: ToolInvocation;
+    activeResponse: ChatState['activeResponse'];
+  }) {
+    if (!activeResponse) return;
+
+    const { toolCallId } = toolInvocation;
+    const { message, partialState } = activeResponse;
+
+    if (!partialState.toolParts) partialState.toolParts = {};
+
+    const existingToolPart = partialState.toolParts[toolCallId] || {};
+    const existingToolInvocation = existingToolPart?.toolInvocation;
+
+    if (existingToolInvocation) {
+      existingToolInvocation.state = toolInvocation.state;
+
+      switch (toolInvocation.state) {
+        case 'partial-call': {
+          const args = existingToolInvocation.args + toolInvocation.args;
+          const { value: partialArgs, state } = await parsePartialJson(args);
+          existingToolInvocation.args =
+            state === 'successful-parse' ? partialArgs : args;
+          break;
+        }
+        case 'call': {
+          existingToolInvocation.args = toolInvocation.args;
+          break;
+        }
+        case 'result': {
+          // @ts-ignore - caused by `result` not existing on ToolCall type
```

Comment:
strange it should have it

Reply:
this is expected due to the fact that the switch narrows down `existingToolInvocation` but not `toolInvocation` even though we set them to be the same earlier in the code

Reply:
gonna leave a better comment here!